### PR TITLE
Add Sparse L1 Attack

### PR DESCRIPTION
**Pull request by zimmerrol**
Implementation of the attack described in [Adversarial Training and Robustness for Multiple Perturbations](https://arxiv.org/pdf/1904.13000.pdf)

For this, the existing code base of the iterative projected gradient descent has been extended so that parameters can be passed from the `as_generator()` method to the calculation of the gradients. This was necessary as the attack sparsifies the original L1 gradient depending on an attack parameter.

### CODE REVIEW

**Comment by jonasrauber** [comment]
Great work, thanks!
Two small requests: could you also add the attack to the docs and could you add a comment why we need to ignore `FutureWarning` (just a comment above the ignore line, like we have it for TensorFlow).
Looking forward to merging this.

**Comment by jonasrauber** [comment]
Why not simply `numpy.percentile` instead of implementing something like that by hand?
And is this code actually doing the last step of the loop of Algorithm 1 in https://arxiv.org/pdf/1904.13000.pdf?

**Comment by FlashTek** [comment]
Oh, you are completely right: I didn't know about `numpy.percentile` but this makes the code way more compact. I will update the PR accordingly.

The last step of Algorithm 1 is performing a projection of the adversarial into the L1 epsilon ball in each iteration step, right? Even though they call it _S1 sphere_ in the paper I am fairly sure that they mean the ball and not only its surface as the referenced [paper ](https://stanford.edu/~jduchi/projects/DuchiShSiCh08.pdf) is focusing on projecting into the ball and not its surface. This should be done by the `_clip_perturbation` function defined in the `L1ClippingMixin`, shouldn't it?

File: foolbox/batch_attacks/iterative_projected_gradient.py [hunk 1]
```diff
@@ -422,6 +457,86 @@ def as_generator(self, a,
...
+            Start the attack from a random point rather than from the
+            original input.
+        return_early : bool
+            Whether an individual gradient descent run should stop as
+            soon as an adversarial is found.
+        """
+
+        assert epsilon > 0
+
+        assert 1.0 <= q <= 99.0, '`q` must be in [0, 100).'
```
Comment (jonasrauber) [inline foolbox/batch_attacks/iterative_projected_gradient.py]: How about `< 100` instead of `<= 99.0` ?

Comment (jonasrauber) [reply foolbox/batch_attacks/iterative_projected_gradient.py]: And actually `== 100` should also be fine (from NumPy's side)

Comment (zimmerrol) [reply foolbox/batch_attacks/iterative_projected_gradient.py]: If one uses `q==100`, then all entries of the gradient will be set to zero because the mask calculated in  [line 213](https://github.com/bethgelab/foolbox/pull/348/files/d0d5cfa178a0bb98620b09a19196b1f4eb9d2662#diff-ad909d83c970d224f1054cb941eb3fd2R213) contains all pixels, won't it?

Comment (jonasrauber) [reply foolbox/batch_attacks/iterative_projected_gradient.py]: True, it's that way round. So please change `<= 99` to `< 100`.

And `>= 1` should be `>= 0`, or not? Wouldn't it be fine to not mask any of them?

**Comment by jonasrauber** [comment]
This implementation was wrong and has been fixed in https://github.com/bethgelab/foolbox/pull/510
